### ISSUE TITLE
angular modules no longer sufficient

### ISSUE THREAD

**Issue by nhhockeyplayer** [angular/angular#46875]
### Which @angular/* package(s) are the source of the bug?

compiler

### Is this a regression?

Yes

### Description

previous versions of angular allowed the creation of modules in a library for me `AdminRootFeatureShellModule`
other source code could consume `AdminRootFeatureShellModule`

private (exports within the library module `AdminRootFeatureShellModule`)
public (exports in an index.ts barrel file popularly now known as the public-api)

previously we were able to build without error with or without barrel file
and one had the option of using barrel files or not
AND the angular library module `AdminRootFeatureShellModule`  is by default a es2015 module anyway usable as-is without error anywhere

So I have chosen to omit using barrel files since they were never needed to begin with since `AdminRootFeatureShellModule` or any modules was an es2015 module anyway

now we receive this error

`modal.component.ts:8:14 - error NG3001: Unsupported private class ModalComponent. This class is visible to consumers via AdminRootFeatureShellModule -> ModalComponent, but is not exported from the top-level library entrypoint.
`

and this has regressed my codebase

Not enough has been documented on the transmissions here
And little is found when searching for the above error codes for a comprehensive explanation and/or solution thats friendly to codebases it just says force yourself into barrel files no if ands or buts. 

Should barrel files have been omitted altogether or are they a band-aid to help angular do its library magic and we are now FORCED to operate them.

I would think we should be at least have the option of NOP'ing out public api barrel file to
index.ts
    `export default {};`

and still be able to operate my library `AdminRootFeatureShellModule.ts` freely since its a es2015 module

why is angular forcing barrel files down out throat

If its functionally required for the special interest of angular thats an issue

This is not backward compatible for es2015 modules in general as im forced to ignore my library module AdminRootFeatureShellModule altogether and rely purely on the barrel file what can be said about whats exported in one as opposed to the other? Conflicted? Not conflicted? Replication of effort?

is this going to be rolled back to re-open the es2015 module freedom?
Or has the angular policy now changed permanently if that policy can be hardcopied here that would be terrific.

If I want to (and its the only way to) export a public-api.ts barrel file it should be optional... no?

If someone can comprehensively speak on this I would appreciate it.

It seems the codebase has gone rogue forcing barrel files forever now.

when we got forward and backward compatible es2015 modules all over the place.

### Please provide a link to a minimal reproduction of the bug

see error its common practice in any simple demo

### Please provide the exception or error you saw

```true
see error its common practice in any simple demo
```

### Anything else?

Not enough has been documented about this change why and the massive maintenance thats now been imposed on very large codebases that have to create massive barrel files to accomodate this

can someone honestly elaborate?

**Comment by jessicajaniuk** [comment]
@nhhockeyplayer Thanks for creating this. Could you provide more information and a reproduction? Which version of angular are you using? What kind of structure do you have? Without information and a repro, it's unclear what you're doing and makes this not very actionable for us.

**Comment by nhhockeyplayer** [comment]
I have a repo (ask to collab if you want in) https://github.com/nhhockeyplayer/meanstacknh

```
nx report

 >  NX   Report complete - copy this into the issue template

   Node : 16.15.1
   OS   : darwin x64
   yarn : 1.22.18
   
   nx : 14.4.3
   @nrwl/angular : 14.4.3
   @nrwl/cypress : 14.4.3
   @nrwl/detox : Not Found
   @nrwl/devkit : 14.4.3
   @nrwl/eslint-plugin-nx : 14.4.3
   @nrwl/express : 14.4.3
   @nrwl/jest : 14.4.3
   @nrwl/js : 14.4.3
   @nrwl/linter : 14.4.3
   @nrwl/nest : 14.4.3
   @nrwl/next : Not Found
   @nrwl/node : 14.4.3
   @nrwl/nx-cloud : 14.2.0
   @nrwl/nx-plugin : Not Found
   @nrwl/react : Not Found
   @nrwl/react-native : Not Found
   @nrwl/schematics : 8.12.11
   @nrwl/storybook : 14.4.3
   @nrwl/web : 14.4.3
   @nrwl/workspace : 14.4.3
   typescript : 4.7.4
   ---------------------------------------
   Community plugins:
   	 @fortawesome/angular-fontawesome: 0.11.1
   	 @ionic/angular: 6.1.14
   	 @ngrx/component: 14.0.2
   	 @ngrx/component-store: 14.0.2
   	 @ngrx/effects: 14.0.2
   	 @ngrx/entity: 14.0.2
   	 @ngrx/router-store: 14.0.2
   	 @ngrx/store: 14.0.2
   	 @nxtend/capacitor: 13.0.0
   	 @nxtend/ionic-angular: 13.1.0
   	 @storybook/angular: 6.5.9
   	 angular-builder-custom-terser-options: 1.0.1
   	 @compodoc/compodoc: 1.1.19
   	 @ngrx/schematics: 14.0.2
   	 @ngrx/store-devtools: 14.0.2
   	 ngx-build-plus: 14.0.0
```

**Comment by nhhockeyplayer** [comment]
Hi Jessica,

repo is buildable as follows

```
yarn install
nx run-many --all --target=build
```

the lib Im interested in augmenting is

`nx build admin-root-feature-shell`

you can change the contents of `index.ts` to `export default {};` to reproduce what I see

I added you as collaborator

feel free to clone

**Comment by nhhockeyplayer** [comment]
> @nhhockeyplayer Can you create this in stackblitz for us instead of a separate repository? I may not be the person that looks into this issue and whomever does on the team will need something clean and repeatable.

sorry im stacked right now the repo is very sufficient

**Comment by nhhockeyplayer** [comment]
this thread speaks to it in short

https://ostack.cn/qa/?qa=248526/

**Comment by alxhub** [comment]
tl;dr: this is working as intended - Angular library compilation requires a single 'barrel' entrypoint that re-exports all publicly reachable Angular symbols.

I doubt this is reproducible in Stackblitz, as I don't think Stackblitz can run the Angular compiler with library options enabled (at least, I've never tried it).

In any case, you're correct that barrel files are a hard requirement for Angular libraries, starting with v9 (with Ivy enabled). v12 was the last version of Angular which could publish the old View Engine format, from v13+ Ivy is the only option.

In Angular, in order to use components from libraries, users import NgModules from those libraries. Our compiler then has to generate imports to those components. To do that, our compiler has to know from _where_ it can import those components - what import specifier should it use?

This problem is much more challenging than it sounds. Our compiler doesn't see the actual library code or even the `package.json` for the library. The only thing we see are the .d.ts files, which may not reflect the reality of how the library is actually packaged (if for example, the JS is bundled but the .d.ts files are not). That means it's not safe to read relative imports in the .d.ts files and assume we can construct a module specifier accordingly.

The best solution turned out to be avoiding this problem in the first place. To do that, Angular uses a simple rule: components accessed via a library NgModule are always imported from the same module specifier that the NgModule itself was. That is, our compiler never tries to create new import paths into libraries - it always uses paths that the user gave us.

To ensure this pattern will always work, when building libraries we enforce that if an NgModule is exported from the library's entrypoint, all components that might be "made available" via that export.

The previous Angular architecture (View Engine) had the same requirement, but was able to re-export such components automatically under the hood, using generated private names. If someone didn't re-export a component exposed by an NgModule from the root, View Engine would do it for them. This workaround was possible because View Engine compiled libraries as part of the application build, whereas Ivy compiles before publishing to NPM. This means Ivy can't generate private names, because they would not be stable across multiple compiled versions.

So in short, yes, this is a new requirement with Ivy. Angular library builds are done against a single entrypoint, and must export all matchable components/directives from that entrypoint. The CLI makes it relatively straightforward to publish packages with many entrypoints, if a particular library wants to expose multiple potential import paths.

More details about the way library code is structured can be found in the [Angular Package Format](https://angular.io/guide/angular-package-format#entrypoints-and-code-splitting) spec.

I hope this explanation makes sense. I'm going to close this as not planned since Angular is working as intended here, but we'll certainly be improving the error messaging / documentation, and feel free to reply/discuss/etc. if something I said above is unclear or confusing.
